File: ps/leetcode_java/src/DuplicateZeros/DuplicateZeros.java
package DuplicateZeros;

import java.util.Arrays;

public class DuplicateZeros {
    // 안해봤지만, 공간 안쓰려면 0은 카운트2, 나머지는 카운트1로 어디까지 마지막 숫자 알아내고, 뒤부터 넣으면 될듯
    public void duplicateZeros(int[] arr) {
        int len = arr.length;
        int[] trr = Arrays.copyOf(arr, len);
        int j = 0;
        for(int i=0; i<len; ++i){
            if(j==len) break;
            if(trr[i] == 0){
                arr[j++] = 0;
                if(j<len) arr[j++] = 0;
            }
            else{
                arr[j++] = trr[i];
            }
        }
    }
}

File: ps/leetcode_cpp/DuplicateZeros.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    // [1,0,2,3,0,4,5,0]
    void duplicateZeros(vector<int>& arr) {
        int len = arr.size();
        int cnt = 0;
        int end = 0;
        for(int i=0; i<len; ++i){
            if(arr[i]==0) cnt += 2;
            else ++cnt;
            if(cnt>=len) {
                end = i;
                break;
            }
        }
        // cnt > len 마지막이 0이고 하나만 넣어야하는 경우
        int j = len - 1;
        if(cnt>len) {
            arr[j--] = 0;
            --end;
        }
        for(int i=end; i>=0; --i){
            if(arr[i]==0){
                arr[j--] = 0;
                arr[j--] = 0;
            }
            else{
                arr[j--] = arr[i];
            }
        }
    }
};


int main(void){

    return 0;
}